src/games/strategy/engine/framework/GameRunner.java

```diff
@@ -666,21 +676,34 @@ public static void joinGame(final GameDescription description, final Messengers
         return;
       }
     }
-    joinGame(description.getPort(), description.getHostedBy().getAddress().getHostAddress(), newClassPath, messengers);
+    joinGame(description.getPort(), description.getHostedBy().getAddress().getHostAddress(),
+        areSameVersion ? "" : newClassPath, messengers);
   }
 
   // newClassPath can be null
   private static void joinGame(final int port, final String hostAddressIP, final String newClassPath,
       final Messengers messengers) {
     final List<String> commands = new ArrayList<>();
-    ProcessRunnerUtil.populateBasicJavaArgs(commands, newClassPath);
-    commands.add("-D" + TRIPLEA_CLIENT_PROPERTY + "=true");
-    commands.add("-D" + TRIPLEA_PORT_PROPERTY + "=" + port);
-    commands.add("-D" + TRIPLEA_HOST_PROPERTY + "=" + hostAddressIP);
-    commands.add("-D" + TRIPLEA_NAME_PROPERTY + "=" + messengers.getMessenger().getLocalNode().getName());
-    final String javaClass = "games.strategy.engine.framework.GameRunner";
-    commands.add(javaClass);
-    ProcessRunnerUtil.exec(commands);
+    final boolean startInNewProcess = !("".equals(newClassPath) && MainFrame.getInstance() != null);
+    String prefix = startInNewProcess ? "-D" : "";
+    if (startInNewProcess) {
+      ProcessRunnerUtil.populateBasicJavaArgs(commands, newClassPath);
+    }
+    commands.add(prefix + TRIPLEA_CLIENT_PROPERTY + "=true");
+    commands.add(prefix + TRIPLEA_PORT_PROPERTY + "=" + port);
+    commands.add(prefix + TRIPLEA_HOST_PROPERTY + "=" + hostAddressIP);
+    commands.add(prefix + TRIPLEA_NAME_PROPERTY + "=" + messengers.getMessenger().getLocalNode().getName());
+    if (startInNewProcess) {
+      final String javaClass = "games.strategy.engine.framework.GameRunner";
+      commands.add(javaClass);
+    }
+    if (startInNewProcess) {
+      ProcessRunnerUtil.exec(commands);
+    } else {
+      MainFrame.getInstance().setVisible(true);
+      handleCommandLineArgs(commands.toArray(new String[commands.size()]), COMMAND_LINE_ARGS, GameMode.SWING_CLIENT);
```

Comment:
The path to having this called is a bit twisted. It also seems like we are trying far too hard to re-use this method. Instead of passing in command line args for that method to do stuff with it, how about we extract the command line arg handling to something that can be re-used in both places, rather than trying to re-use `handleCommandLineargs`?